services/shard_precreation/service.go

```diff
@@ -0,0 +1,122 @@
+package shard_precreation
+
+import (
+	"log"
+	"os"
+	"sync"
+	"time"
+
+	"github.com/influxdb/influxdb/meta"
+)
+
+type Service struct {
+	checkInterval time.Duration
+	advancePeriod time.Duration
+
+	Logger *log.Logger
+
+	done chan struct{}
+	wg   sync.WaitGroup
+
+	MetaStore interface {
+		IsLeader() bool
+		VisitRetentionPolicies(f func(d meta.DatabaseInfo, r meta.RetentionPolicyInfo))
+		ShardGroups(database, policy string) ([]meta.ShardGroupInfo, error)
+		CreateShardGroupIfNotExists(database, policy string, timestamp time.Time) (*meta.ShardGroupInfo, error)
+	}
+}
+
+// NewService returns an instance of the Graphite service.
+func NewService(c Config) (*Service, error) {
+	s := Service{
+		checkInterval: time.Duration(c.CheckInterval),
+		advancePeriod: time.Duration(c.AdvancePeriod),
+		Logger:        log.New(os.Stderr, "[shard-precreation] ", log.LstdFlags),
+	}
+
+	return &s, nil
+}
+
+// SetLogger sets the internal logger to the logger passed in.
+func (s *Service) SetLogger(l *log.Logger) {
+	s.Logger = l
+}
+
+// Open starts the shard precreation service.
+func (s *Service) Open() error {
+	if s.done != nil {
+		return nil
+	}
+
+	s.done = make(chan struct{})
+
+	s.wg.Add(1)
+	go s.runPrecreation()
+	return nil
+}
+
+// Close stops the shard precreation service.
+func (s *Service) Close() error {
+	if s.done == nil {
+		return nil
+	}
+
+	close(s.done)
+	s.wg.Wait()
+	s.done = nil
+
+	return nil
+}
+
+// runPrecreation continually checks if shards need precreation.
+func (s *Service) runPrecreation() {
+	defer s.wg.Done()
+
+	for {
+		select {
+		case <-time.After(s.checkInterval):
```

Comment:
It could be, and that would be fine, but we don't need that fine a control over when shard precreation runs. A small jitter is a not big deal, since iterations of this loop only happen on the order of minutes. It seems in this case the less code is nicer.